builder/builder.go

```diff
@@ -29,27 +31,100 @@ type Builder struct {
 	privKey    *ecdsa.PrivateKey
 
 	bc       *blockchain.Blockchain
+	pool     *mempool.Pool
 	vm       vm.VM
 	newHeads *feed.Feed[*core.Header]
 	log      utils.Logger
+
+	pendingLock  stdsync.Mutex
+	pendingBlock blockchain.Pending
+	headState    core.StateReader
+	headCloser   blockchain.StateCloser
 }
 
-func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchain, builderVM vm.VM, log utils.Logger) *Builder {
+func New(privKey *ecdsa.PrivateKey, ownAddr *felt.Felt, bc *blockchain.Blockchain, pool *mempool.Pool, builderVM vm.VM, log utils.Logger) *Builder {
 	return &Builder{
 		ownAddress: *ownAddr,
 		privKey:    privKey,
 
 		bc:       bc,
+		pool:     pool,
 		vm:       builderVM,
 		newHeads: feed.New[*core.Header](),
+		log:      log,
 	}
 }
 
 func (b *Builder) Run(ctx context.Context) error {
-	<-ctx.Done()
+	if err := b.initPendingBlock(); err != nil {
+		return err
+	}
+	defer b.clearPending()
+
+	go b.listenPool(ctx)
+
+	const blockTime = 5 * time.Second
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(blockTime):
+			if err := b.Finalise(); err != nil {
```

Comment:
nit: do you think it would be clearer to have Finalise return nil on a success, and then call b.initPendingBlock() after?